File: progect2/progect2/Core/TSDF/SpatialHashTable.swift
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

//
// SpatialHashTable.swift
// Aether3D
//
// Sparse voxel storage using spatial hashing with separated metadata

import Foundation

struct HashEntry: Sendable {
    var key: BlockIndex    // Int32 x, Int32 y, Int32 z — 12 bytes
    var blockPoolIndex: Int32  // Index into VoxelBlockPool, -1 if empty — 4 bytes
}

/// Sparse voxel storage using spatial hashing with separated metadata
public struct SpatialHashTable: Sendable {
    private var entries: ContiguousArray<HashEntry>
    private var stableKeyList: ContiguousArray<BlockIndex>  // Deterministic iteration
    private var pool: VoxelBlockPool
    public private(set) var count: Int = 0
    public var tableCapacity: Int { entries.count }

    /// Initial table size: 2^16 = 65,536
    public init(
        initialSize: Int = TSDFConstants.hashTableInitialSize,
        poolCapacity: Int = TSDFConstants.maxTotalVoxelBlocks
    ) {
        entries = ContiguousArray(repeating: HashEntry(key: BlockIndex(0,0,0), blockPoolIndex: -1), count: initialSize)
        stableKeyList = ContiguousArray()
        pool = VoxelBlockPool(capacity: poolCapacity)
        count = 0
    }

    public mutating func insertOrGet(key: BlockIndex, voxelSize: Float) -> Int? {
        let hash = key.niessnerHash(tableSize: entries.count)
        var probe = hash
        
        // Linear probing
        for _ in 0..<TSDFConstants.hashMaxProbeLength {
            let entry = entries[probe]
            
            if entry.blockPoolIndex == -1 {
                // Empty slot - allocate new block
                guard let poolIndex = pool.allocate(voxelSize: voxelSize) else { return nil }
                entries[probe] = HashEntry(key: key, blockPoolIndex: Int32(poolIndex))
                stableKeyList.append(key)
                count += 1
                return poolIndex
            }
            
            if entry.key == key {
                // Found existing block
                return Int(entry.blockPoolIndex)
            }
            
            probe = (probe + 1) % entries.count
        }
        
        return nil  // Max probe length exceeded
    }

    public func lookup(key: BlockIndex) -> Int? {
        let hash = key.niessnerHash(tableSize: entries.count)
        var probe = hash
        
        for _ in 0..<TSDFConstants.hashMaxProbeLength {
            let entry = entries[probe]
            
            if entry.blockPoolIndex == -1 {
                return nil  // Empty slot, not found
            }
            
            if entry.key == key {
                return Int(entry.blockPoolIndex)
            }
            
            probe = (probe + 1) % entries.count
        }
        
        return nil
    }

    public mutating func remove(key: BlockIndex) {
        // BUG-5: Backward-shift deletion to preserve linear probing chain
        let h = key.niessnerHash(tableSize: entries.count)
        var probe = h
        
        for _ in 0..<TSDFConstants.hashMaxProbeLength {
            let entry = entries[probe]
            if entry.blockPoolIndex == -1 { return }
            if entry.key == key {
                pool.deallocate(index: Int(entry.blockPoolIndex))
                stableKeyList.removeAll { $0 == key }
                var empty = probe
                var j = (probe + 1) % entries.count
                while entries[j].blockPoolIndex >= 0 {
                    let ideal = entries[j].key.niessnerHash(tableSize: entries.count)
                    if (empty <= j) ? (ideal <= empty || ideal > j) : (ideal <= empty && ideal > j) {
                        entries[empty] = entries[j]
                        empty = j
                    }
                    j = (j + 1) % entries.count
                }
                entries[empty] = HashEntry(key: BlockIndex(0,0,0), blockPoolIndex: -1)
                count -= 1
                return
            }
            probe = (probe + 1) % entries.count
        }
    }

    /// Load factor check — rehash at 0.7
    /// Guardrail #24: Hash load factor monitoring
    public var loadFactor: Float { Float(count) / Float(entries.count) }
    
    /// Guardrail #24: Check if load factor exceeds warning threshold
    public var loadFactorWarning: Bool {
        loadFactor >= 0.6  // Warn at 0.6, rehash at 0.7
    }

    public mutating func rehashIfNeeded() {
        // Guardrail #24: Proactive rehash at 0.7 — only remap key→poolIndex, do not re-allocate pool blocks (BUG-4 fix)
        guard loadFactor >= TSDFConstants.hashTableMaxLoadFactor else { return }
        
        let oldMappings: [(BlockIndex, Int32)] = entries.compactMap { entry in
            entry.blockPoolIndex >= 0 ? (entry.key, entry.blockPoolIndex) : nil
        }
        let newSize = entries.count * 2
        entries = ContiguousArray(repeating: HashEntry(key: BlockIndex(0,0,0), blockPoolIndex: -1), count: newSize)
        stableKeyList = ContiguousArray()
        count = 0
        
        for (key, poolIndex) in oldMappings {
            let h = key.niessnerHash(tableSize: entries.count)
            var probe = h
            for _ in 0..<TSDFConstants.hashMaxProbeLength {
                if entries[probe].blockPoolIndex == -1 {
                    entries[probe] = HashEntry(key: key, blockPoolIndex: poolIndex)
                    stableKeyList.append(key)
                    count += 1
                    break
                }
                probe = (probe + 1) % entries.count
            }
        }
    }

    // ── Exposed for TSDFVolume and MetalTSDFIntegrator ──

    /// VoxelBlockAccessor for TSDFIntegrationBackend protocol dispatch.
    public var voxelAccessor: VoxelBlockAccessor { pool.accessor }

    /// Stable base address for Metal MTLBuffer(bytesNoCopy:) binding.
    public var voxelBaseAddress: UnsafeMutableRawPointer { pool.baseAddress }
    public var voxelByteCount: Int { pool.byteCount }

    /// Read a voxel block by pool index (for MarchingCubes).
    public func readBlock(at poolIndex: Int) -> VoxelBlock { pool.accessor.readBlock(at: poolIndex) }
    
    /// Iterate through all active blocks in the hash table.
    /// Used by MarchingCubes and memory pressure handling.
    public func forEachBlock(_ block: (BlockIndex, Int, VoxelBlock) -> Void) {
        for key in stableKeyList {
            if let poolIndex = lookup(key: key) {
                let voxelBlock = readBlock(at: poolIndex)
                block(key, poolIndex, voxelBlock)
            }
        }
    }
    
    /// Get all block indices and their pool indices.
    /// Returns array of (BlockIndex, poolIndex) tuples.
    public func getAllBlocks() -> [(BlockIndex, Int)] {
        var result: [(BlockIndex, Int)] = []
        for key in stableKeyList {
            if let poolIndex = lookup(key: key) {
                result.append((key, poolIndex))
            }
        }
        return result
    }
    
    /// Mutate a block at the given pool index.
    /// Used to update meshGeneration after meshing.
    public mutating func updateBlock(at poolIndex: Int, _ updater: (inout VoxelBlock) -> Void) {
        var block = pool.accessor.readBlock(at: poolIndex)
        updater(&block)
        pool.accessor.writeBlock(at: poolIndex, block)
    }
}

File: progect2/progect2/aether_cpp/src/tsdf/spatial_hash_table.cpp
// SPDX-License-Identifier: LicenseRef-Aether3D-Proprietary
// Copyright (c) 2024-2026 Aether3D. All rights reserved.

#include "aether/tsdf/spatial_hash_table.h"
#include <cstdlib>
#include <cstring>

namespace aether {
namespace tsdf {

namespace {

inline int round_up_pow2(int value) {
    int v = (value <= 1) ? 1 : value;
    --v;
    v |= v >> 1;
    v |= v >> 2;
    v |= v >> 4;
    v |= v >> 8;
    v |= v >> 16;
    ++v;
    return v;
}

inline int probe_index(int start, int probe, int capacity) {
    return (start + probe) & (capacity - 1);
}

}  // namespace

void SpatialHashTable::init(int table_size, int stable_key_capacity) {
    reset();
    capacity_ = round_up_pow2(table_size > 0 ? table_size : HASH_TABLE_INITIAL_SIZE);
    entries_ = static_cast<Entry*>(std::malloc(static_cast<size_t>(capacity_) * sizeof(Entry)));
    occupied_ = static_cast<uint8_t*>(std::calloc(static_cast<size_t>(capacity_), sizeof(uint8_t)));
    stable_key_capacity_ = stable_key_capacity > 0 ? stable_key_capacity : MAX_TOTAL_VOXEL_BLOCKS;
    stable_keys_ = static_cast<BlockIndex*>(std::malloc(static_cast<size_t>(stable_key_capacity_) * sizeof(BlockIndex)));
    if (!entries_ || !occupied_ || !stable_keys_) {
        std::free(entries_);
        std::free(occupied_);
        std::free(stable_keys_);
        entries_ = nullptr;
        occupied_ = nullptr;
        stable_keys_ = nullptr;
        capacity_ = 0;
        stable_key_capacity_ = 0;
        stable_key_count_ = 0;
        size_ = 0;
        return;
    }
    stable_key_count_ = 0;
    size_ = 0;
    for (int i = 0; i < capacity_; ++i) {
        entries_[static_cast<size_t>(i)] = Entry{};
    }
}

int SpatialHashTable::find_slot(const BlockIndex& bi) const {
    if (!entries_ || !occupied_ || capacity_ <= 0) return -1;
    const int start = bi.niessner_hash(capacity_);
    const int max_probe = capacity_ < HASH_MAX_PROBE_LENGTH ? capacity_ : HASH_MAX_PROBE_LENGTH;
    for (int probe = 0; probe < max_probe; ++probe) {
        const int idx = probe_index(start, probe, capacity_);
        if (!occupied_[static_cast<size_t>(idx)]) return -1;
        if (entries_[static_cast<size_t>(idx)].key == bi) return idx;
    }
    return -1;
}

int SpatialHashTable::find_insert_slot(const BlockIndex& bi) const {
    if (!entries_ || !occupied_ || capacity_ <= 0) return -1;
    const int start = bi.niessner_hash(capacity_);
    const int max_probe = capacity_ < HASH_MAX_PROBE_LENGTH ? capacity_ : HASH_MAX_PROBE_LENGTH;
    for (int probe = 0; probe < max_probe; ++probe) {
        const int idx = probe_index(start, probe, capacity_);
        if (!occupied_[static_cast<size_t>(idx)] ||
            entries_[static_cast<size_t>(idx)].key == bi) {
            return idx;
        }
    }
    return -1;
}

int SpatialHashTable::find(const BlockIndex& bi) const {
    return find_slot(bi);
}

void SpatialHashTable::append_stable_key(const BlockIndex& bi) {
    if (stable_key_count_ < stable_key_capacity_) {
        stable_keys_[static_cast<size_t>(stable_key_count_++)] = bi;
        return;
    }
    const int new_capacity = stable_key_capacity_ > 0 ? stable_key_capacity_ * 2 : 64;
    BlockIndex* grown = static_cast<BlockIndex*>(
        std::realloc(stable_keys_, static_cast<size_t>(new_capacity) * sizeof(BlockIndex)));
    if (!grown) return;
    stable_keys_ = grown;
    stable_key_capacity_ = new_capacity;
    stable_keys_[static_cast<size_t>(stable_key_count_++)] = bi;
}

void SpatialHashTable::erase_stable_key(const BlockIndex& bi) {
    for (int i = 0; i < stable_key_count_; ++i) {
        if (stable_keys_[static_cast<size_t>(i)] == bi) {
            stable_keys_[static_cast<size_t>(i)] = stable_keys_[static_cast<size_t>(stable_key_count_ - 1)];
            --stable_key_count_;
            return;
        }
    }
}

float SpatialHashTable::load_factor() const {
    if (capacity_ <= 0) return 0.0f;
    return static_cast<float>(size_) / static_cast<float>(capacity_);
}

void SpatialHashTable::rehash_to(int new_capacity) {
    new_capacity = round_up_pow2(new_capacity);
    if (new_capacity <= capacity_ || size_ == 0) return;
    Entry* new_entries = static_cast<Entry*>(std::malloc(static_cast<size_t>(new_capacity) * sizeof(Entry)));
    uint8_t* new_occupied = static_cast<uint8_t*>(std::calloc(static_cast<size_t>(new_capacity), sizeof(uint8_t)));
    if (!new_entries || !new_occupied) {
        std::free(new_entries);
        std::free(new_occupied);
        return;
    }
    for (int i = 0; i < new_capacity; ++i) {
        new_entries[static_cast<size_t>(i)] = Entry{};
    }
    for (int i = 0; i < capacity_; ++i) {
        if (!occupied_[static_cast<size_t>(i)]) continue;
        const Entry entry = entries_[static_cast<size_t>(i)];
        const int start = entry.key.niessner_hash(new_capacity);
        const int max_probe = new_capacity < HASH_MAX_PROBE_LENGTH ? new_capacity : HASH_MAX_PROBE_LENGTH;
        for (int probe = 0; probe < max_probe; ++probe) {
            const int idx = probe_index(start, probe, new_capacity);
            if (!new_occupied[static_cast<size_t>(idx)]) {
                new_entries[static_cast<size_t>(idx)] = entry;
                new_occupied[static_cast<size_t>(idx)] = 1;
                break;
            }
        }
    }
    std::free(entries_);
    std::free(occupied_);
    entries_ = new_entries;
    occupied_ = new_occupied;
    capacity_ = new_capacity;
}

void SpatialHashTable::rehash_if_needed() {
    if (capacity_ <= 0) return;
    if (load_factor() >= HASH_TABLE_MAX_LOAD_FACTOR) {
        rehash_to(capacity_ * 2);
    }
}

int SpatialHashTable::insert_or_get(const BlockIndex& bi, int value) {
    if (!entries_ || !occupied_ || capacity_ <= 0) return -1;
    rehash_if_needed();
    const int start = bi.niessner_hash(capacity_);
    const int max_probe = capacity_ < HASH_MAX_PROBE_LENGTH ? capacity_ : HASH_MAX_PROBE_LENGTH;
    for (int probe = 0; probe < max_probe; ++probe) {
        const int idx = probe_index(start, probe, capacity_);
        if (!occupied_[static_cast<size_t>(idx)]) {
            entries_[static_cast<size_t>(idx)].key = bi;
            entries_[static_cast<size_t>(idx)].value = value;
            occupied_[static_cast<size_t>(idx)] = 1;
            ++size_;
            append_stable_key(bi);
            return value;
        }
        if (entries_[static_cast<size_t>(idx)].key == bi) {
            return entries_[static_cast<size_t>(idx)].value;
        }
    }
    return -1;
}

int SpatialHashTable::lookup(const BlockIndex& bi) const {
    const int slot = find_slot(bi);
    if (slot < 0) return -1;
    return entries_[static_cast<size_t>(slot)].value;
}

int SpatialHashTable::insert(const BlockIndex& bi) {
    if (!entries_ || !occupied_ || capacity_ <= 0) return -1;
    rehash_if_needed();
    const int start = bi.niessner_hash(capacity_);
    const int max_probe = capacity_ < HASH_MAX_PROBE_LENGTH ? capacity_ : HASH_MAX_PROBE_LENGTH;
    for (int probe = 0; probe < max_probe; ++probe) {
        const int idx = probe_index(start, probe, capacity_);
        if (!occupied_[static_cast<size_t>(idx)]) {
            entries_[static_cast<size_t>(idx)].key = bi;
            entries_[static_cast<size_t>(idx)].value = idx;
            occupied_[static_cast<size_t>(idx)] = 1;
            ++size_;
            append_stable_key(bi);
            return idx;
        }
        if (entries_[static_cast<size_t>(idx)].key == bi) {
            return idx;
        }
    }
    return -1;
}

bool SpatialHashTable::remove(const BlockIndex& bi) {
    if (!entries_ || !occupied_ || capacity_ <= 0) return false;
    int empty = find_slot(bi);
    if (empty < 0) return false;

    erase_stable_key(bi);

    int cursor = probe_index(empty, 1, capacity_);
    while (occupied_[static_cast<size_t>(cursor)]) {
        const int ideal = entries_[static_cast<size_t>(cursor)].key.niessner_hash(capacity_);
        const bool move =
            (cursor > empty)
                ? (ideal <= empty || ideal > cursor)
                : (ideal <= empty && ideal > cursor);
        if (move) {
            entries_[static_cast<size_t>(empty)] = entries_[static_cast<size_t>(cursor)];
            occupied_[static_cast<size_t>(empty)] = 1;
            empty = cursor;
        }
        cursor = probe_index(cursor, 1, capacity_);
    }

    entries_[static_cast<size_t>(empty)] = Entry{};
    occupied_[static_cast<size_t>(empty)] = 0;
    --size_;
    return true;
}

void SpatialHashTable::reset() {
    std::free(entries_);
    std::free(occupied_);
    std::free(stable_keys_);
    entries_ = nullptr;
    occupied_ = nullptr;
    stable_keys_ = nullptr;
    stable_key_capacity_ = 0;
    stable_key_count_ = 0;
    capacity_ = 0;
    size_ = 0;
}

}  // namespace tsdf
}  // namespace aether
